Improve logging

tobim opened the pull request:
This PR attempts to make the VASTs log more consistent.

Tasks:
- [x] Let `VAST_{ERROR,WARNING,INFO,DEBUG}` reject `const char*` as its first parameter
- [x] Add `VAST_{ERROR,WARNING,INFO,DEBUG}_` to bypass this check.
- [x] Convert messages to *subject verb object* style
- [x] Reserve `VAST_INFO` to the main system actors (SINK, SOURCE, IMPORTER, ARCHIVE, INDEX, EXPORTER)
- [x] Add logging guidelines to`CONTRIBUTING.md`

mavam commented on libvast/src/system/node_command.cpp:
```diff
@@ -99,7 +99,7 @@ node_command::connect_to_node(scoped_actor& self,
     return make_error(sec::invalid_argument, std::move(err));
   }
-  VAST_INFO(&self, "connect to remote node:", id);
+  VAST_DEBUG(&self, "connect to remote node:", id);
```
s/connect/connects/

mavam replied:
Also, `self` should work here as well without the ampersand.

mavam commented:
Perhaps squash some of the commits, @tobim? Seems like there are some repetitions.

mavam commented on CMakeLists.txt:
```diff
@@ -37,6 +37,10 @@ list(GET version_numbers 0 VERSION_MAJOR)
 set(VERSION_MAJ_MIN "${VERSION_MAJOR}.${VERSION_MINOR}")
 
+if (NOT VAST_LOG_LEVEL)
+  set(VAST_LOG_LEVEL "-1")
```
Would you mind elaborating why you set the log level to a string an not a numeric value?

tobim replied:
cmake does not have an integral type. There is no semantic difference between `-1` and `"-1"`. I added the quotation marks to not confuse the reader about this.

mavam commented on libvast/src/column_index.cpp:
```diff
@@ -95,20 +95,20 @@ caf::error column_index::init() {
     detail::value_index_inspect_helper tmp{index_type_, idx_};
     if (auto err = load(sys_, filename_, last_flush_, tmp)) {
-      VAST_ERROR("unable to load value index from disk", sys_.render(err));
+      VAST_ERROR(this, "unable to load value index from disk", sys_.render(err));
```
Seems like the sentence doesn't quite read properly: `this, "failed to load..."`?

mavam commented on libvast/src/column_index.cpp:
```diff
@@ -118,7 +118,7 @@ caf::error column_index::flush_to_disk() {
   if (offset == last_flush_)
     return caf::none;
-  VAST_DEBUG("flush index (" << (offset - last_flush_) << '/' << offset,
+  VAST_DEBUG(this, "flush index (" << (offset - last_flush_) << '/' << offset,
```
`flush` should read `flushes` or `flushed`.

mavam commented on libvast/src/meta_index.cpp:
```diff
@@ -49,11 +49,12 @@ operator[](const uuid& partition) const {
 void meta_index::add(const uuid& partition,
                      const const_table_slice_handle& slice) {
   auto& rng = partitions_[partition].range;
+  VAST_DEBUG(this, "meta_index got table slice:", slice);
```
This doesn't read like a proper sentence.

tobim replied:
done

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -57,7 +57,7 @@ segment_store::~segment_store() {
 caf::error segment_store::put(const_table_slice_handle xs) {
   VAST_TRACE(VAST_ARG(xs));
-  VAST_DEBUG("adding a table slice");
+  VAST_DEBUG(this, "adding a table slice");
```
`adds`?

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -78,8 +78,8 @@ caf::error segment_store::flush() {
-  VAST_DEBUG("wrote new segment to", filename.trim(-3));
-  VAST_DEBUG("saving segment meta data");
+  VAST_DEBUG(this, "wrote new segment to", filename.trim(-3));
+  VAST_DEBUG(this, "saving segment meta data");
```
`saves`?

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -88,7 +88,7 @@ segment_store::get(const ids& xs) {
   // Collect candidate segments by seeking through the ID set and
   // probing each ID interval.
-  VAST_DEBUG("getting table slices with ids");
+  VAST_DEBUG(this, "getting table slices with ids");
```
s/getting/retrieves/ ?

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
   // Process candidates in reverse order for maximum LRU cache hits.
   std::vector<const_table_slice_handle> result;
-  VAST_DEBUG("processing", candidates.size(), "candidates");
+  VAST_DEBUG(this, "processing", candidates.size(), "candidates");
```
`processes`

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
     caf::expected<std::vector<const_table_slice_handle>> slices{caf::no_error};
     if (id == builder_.id()) {
-      VAST_DEBUG("looking into builder");
+      VAST_DEBUG(this, "looking into builder");
```
`looks`

mavam commented on libvast/src/segment_store.cpp:
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
         auto fname = segment_path() / to_string(id);
         if (auto err = load(sys_, fname, seg_ptr)) {
-          VAST_ERROR("unable to load segment:", sys_.render(err));
+          VAST_ERROR(this, "unable to load segment:", sys_.render(err));
```
`failed to load...`

mavam replied:
What about this one, @tobim?

(By the way, github has now a "resolved" feature where you can check off each addressed review comment.)

mavam commented on libvast/src/system/exporter.cpp:
```diff
@@ -42,9 +42,11 @@ namespace system {
 
 void ship_results(stateful_actor<exporter_state>* self) {
-  if (self->state.results.empty() || self->state.stats.requested == 0)
+  VAST_TRACE("");
```
What's the benefit of an empty string as trace log entry?

tobim replied:
You are right to point out that messages like this are a net negative if the log format is not chosen appropriately. The current console logger will just show `ENTER` and `EXIT` messages. But the file logger writes the filename, line and function name, which is mostly what we want from the traces. Also, there is some precedent in the code base for doing it that way.

I would like to extend the logger to support different levels for file and console output. Then the empty trace messages won't be a problem any more.

mavam replied:
> Also, there is some precedent in the code base for doing it that way.

@Neverlord, would you mind filling in the rationale for those? I think these come from you.

> I would like to extend the logger to support different levels for file and console output. Then the empty trace messages won't be a problem any more.

Makes sense. That seems to be a straight shot in the CAF codebase.

Neverlord replied:
This emits valuable information for following program flow with ENTER and EXIT messages for the function. This is used whenever no function arguments are available (other than `self` or the implicit `this` pointer) or are omitted for some other reason (not reasonably printable or something along those lines).

mavam commented on libvast/src/system/node.cpp:
```diff
@@ -306,7 +306,7 @@ caf::behavior node(node_ptr self, std::string id, path dir) {
       } else {
         auto e = make_error(ec::unspecified, "invalid command", cmd);
-        VAST_INFO(self, self->system().render(e));
+        VAST_WARNING(self, self->system().render(e));
```
Maybe add a little message here, like `failed to parse command`.

mavam commented on libvast/src/system/node_command.cpp:
```diff
@@ -99,7 +99,7 @@ node_command::connect_to_node(scoped_actor& self,
     return make_error(sec::invalid_argument, std::move(err));
   }
-  VAST_INFO("connect to remote node:", id);
+  VAST_DEBUG(&self, "connect to remote node:", id);
```
`connects`

mavam commented on libvast/src/system/node_command.cpp:
```diff
@@ -109,7 +109,7 @@ node_command::connect_to_node(scoped_actor& self,
   if (node_endpoint.host.empty())
     node_endpoint.host = "127.0.0.1";
-  VAST_INFO("connecting to", node_endpoint.host << ':' << node_endpoint.port);
+  VAST_INFO(&self, "connecting to", node_endpoint.host << ':' << node_endpoint.port);
```
`connects`

mavam commented on libvast/src/system/partition.cpp:
```diff
@@ -46,7 +46,7 @@ partition::partition(caf::actor_system& sys, const path& base_dir, uuid id,
   if (exists(dir_)) {
     if (auto err = load(sys_, meta_file(), meta_data_)) {
-      VAST_ERROR("unable to read partition meta data:", sys_.render(err));
+      VAST_ERROR(this, "unable to read meta data:", sys_.render(err));
```
s/unable/failed/

mavam commented on libvast/src/system/reader_command_base.cpp:
```diff
@@ -69,20 +69,20 @@ int reader_command_base::run_impl(caf::actor_system& sys,
         stop = true;
       } else {
-        VAST_DEBUG("connecting source to importer");
+        VAST_DEBUG(this, "connecting to importer");
```
`connects`

mavam commented on libvast/src/system/writer_command_base.cpp:
```diff
@@ -47,7 +47,7 @@ int writer_command_base::run_impl(caf::actor_system& sys,
   });
   // Spawn a sink.
-  VAST_DEBUG("spawning sink with parameters:", deep_to_string(options));
+  VAST_DEBUG(this, "spawning sink with parameters:", deep_to_string(options));
```
`spawns`

mavam commented on libvast/src/system/writer_command_base.cpp:
```diff
@@ -70,14 +70,14 @@ int writer_command_base::run_impl(caf::actor_system& sys,
   auto max_events = get_or<uint64_t>(options, "events", 0u);
   args += make_message("-e", std::to_string(max_events));
-  VAST_DEBUG("spawning exporter with parameters:", to_string(args));
+  VAST_DEBUG(this, "spawning exporter with parameters:", to_string(args));
```
`spawns`

mavam commented on libvast/src/table_index.cpp:
```diff
@@ -261,7 +261,7 @@ caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
     return lookup_impl(pred, dx, x);
   }
-  VAST_WARNING("unsupported attribute:", ex.attr);
+  VAST_WARNING(this, "unsupported attribute:", ex.attr);
```
`got unsupported...`

mavam commented on libvast/src/table_index.cpp:
```diff
@@ -280,7 +280,7 @@ caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
   auto index = r.flat_index_at(dx.offset);
   if (!index) {
-    VAST_DEBUG("invalid offset for record type", dx.type);
+    VAST_DEBUG(this, "invalid offset for record type", dx.type);
```
`got invalid offset...`

mavam commented on libvast/src/table_slice.cpp:
```diff
@@ -76,7 +76,8 @@ table_slice_ptr table_slice::make_ptr(record_type layout,
   if (!caf::holds_alternative<generic_fun>(val)) {
-    VAST_ERROR("no factory function stored for implementation key", impl);
+    VAST_ERROR_("table_slice",
+                  "has no factory function for implementation key", impl);
```
Style: indentation.

mavam commented on libvast/vast/logger.hpp:
```diff
@@ -48,34 +50,93 @@
+  static_assert(
+    !std::is_same_v<const char*, std::remove_reference<T>>,
+    "const char* is not allowed for the first argument in a logging statement"
+    "supply a component or use `VAST_[ERROR|WARNING|INFO|DEBUG]_ANON` instead");
```
Is `_ANON` now simply a `_` suffix?

tobim replied:
Yes, I missed updating that error message.
The suffix transitioned from `_FF` to `_ANON` to `_NS` (for *no subject*). But all of them suggest a wrong meaning at some use-sites. Maybe `_CS` for *custom subject*?

tobim replied:
As per our discussion, reverted to `_ANON`.

mavam commented on libvast/vast/logger.hpp:
```diff
@@ -48,34 +50,93 @@
+    "const char* is not allowed for the first argument in a logging statement"
+    "supply a component or use `VAST_[ERROR|WARNING|INFO|DEBUG]_ANON` instead");
+
+  using Type = std::remove_pointer_t<T>;
```
Capitalized type aliases conflict with our coding style.

tobim commented:
I think all change requests are now addressed in 03878594a0c333766a422bcfec07a3f1b671ddd2.

mavam commented:
I still see several unaddressed comments, e.g., `...(this, "getting table slice...`.

Per my earlier request, would you mind squashing the commits a bit? There are two identically named commits in there, which seems like they should be one commit.

tobim commented:
Thanks for pointing that out again, I updated this PR once more and hope that I did not miss any open requests this time. Also, I squashed the two "Tune log levels" commits into one.

mavam commented:
Great work, this is the foundation for more usability and easier debugging in the future. Can't wait to sprinkle the new macros over my code. 😉
